Column grouping sidebar

mattrunyon opened the pull request:
Fixes #962

I strayed from the spec on creating groups and instead opted for a button that is only active when a user has selected items. The selected items will be moved into a new group instead. This removes a bunch of places that would otherwise need a special case just for handling the new group item.

Items left todo

- [x] CSS cleanup
- [x] Fix TS (mostly some templated types that need to be piped all the way through)
- [x] Change selections to be name based instead of index so groups can be selected
- [x] Update create group to allow creation with selected groups
- [x] Update create group to move non-adjacent items next to each other
- [x] Finish multi-select handling
> - [x] Drag and drop
> - [x] Move up/down/top/bottom
- [x] Write a comprehensive list of behaviors to test (there's a lot in the unit tests)
- [x] Should probably write unit tests

dsmmcken commented:
> I strayed from the spec on creating groups and instead opted for a button that is only active when a user has selected items. The selected items will be moved into a new group instead. 

Acknowledging approval of this change. Not ideal, but acceptable trade-off in terms of development effort.

mattrunyon commented:
I do think there's some benefit to the change from usability for the expediency of adding a group. Spec as written would require click add group, name group, select columns you want in group, drag into group. This at least just requires select columns, click create group, name group. Skips the need for the drag on initial creation if you know the columns you want

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+
+export function SortableTree<T = undefined>({
+  collapsible = false,
+  defaultItems = [] as TreeItem<T>[],
```
Define this array outside as a constant.
Like this as a default arg, it could cause an infinite loop since the `[]` will not equal twice through two renders (specifically that `useEffect` using `defaultItems` in it's dependency array below).

mattrunyon replied:
I think it should just be required. This was code from dnd-kit examples

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+        distance: 5,
+      },
+    })
+    // useSensor(KeyboardSensor, {
+    //   coordinateGetter,
+    // })
```
Unused
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+                    : undefined
+                }
+                onRemove={
+                  removable && onRemove ? () => handleRemove(id) : undefined
```
`handleRemove` undefined - I think you mean `onRemove(item)`.

Could also set `onRemove` to a default function (defined outside as a const) and then this could just be `removable ? onRemove(item) : undefined`

mattrunyon replied:
I'm removing the removable and collapsible code from this. It was part of the example and we aren't using it. Should be easy enough to add back later if desired

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+                value={activeId.toString()}
+                indentationWidth={indentationWidth}
+                renderItem={renderItem}
+                item={activeItem}
```
Type error

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+    setActiveId(newActiveId);
+    setOverId(newActiveId);
+
+    document.body.style.setProperty('cursor', 'grabbing');
```
Not sure if there's any differences between this and doing it on `document.documentElement` like `Grid` does. In any case, this works.

mattrunyon replied:
Probably not unless you render content outside the body

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+
+  function handleCollapse(id: UniqueIdentifier) {
+    setItems(oldItems =>
+      setProperty(oldItems, id, 'collapsed', value => !value)
```
Need to clean up typing on `setProperty`

mattrunyon replied:
Not needed any more since I'm removing collapse. Not using that functionality

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+  }
+}
+
+const adjustTranslate: Modifier = ({ transform }) => {
```
I'd prefer `function adjustTranslate`

mattrunyon replied:
It's typed this way because `Modifier` is a function type. Changing would have to be `function adjustTranslate({ transform }: Parameters<Modifier>[0]): ReturnType<Modifier> { ... }` which is much more fragile

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTree.tsx:
```diff
@@ -0,0 +1,350 @@
+    setOverId(over?.id ?? null);
+  }
+
+  function handleDragEnd({ active, over }: DragEndEvent) {
```
Not sure I like all these functions declared after the return; should also be using `useCallback` with a proper dependencies array.

mattrunyon replied:
Ya no idea why the example did it that way. I'll clean up

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTreeItem.tsx:
```diff
@@ -0,0 +1,54 @@
+  id: UniqueIdentifier;
+}
+
+const animateLayoutChanges: AnimateLayoutChanges = ({
```
Declare as `function`?

mattrunyon replied:
Same as before. Not as clean in TS to declare as `function` since `AnimateLayoutChanges` is a function type

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/SortableTreeItem.tsx:
```diff
@@ -0,0 +1,54 @@
+    id,
+    animateLayoutChanges,
+  });
+  const style: CSSProperties = {
+    transform: CSS.Translate.toString(transform),
+    transition,
+  };
```
I'd probably memoize this just to ensure no unnecessary re-rendering of the child.

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  &.clone {
+    // display: inline-block;
```
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+          height: 12px;
+          border-radius: 50%;
+          border: 1px solid #2389ff;
+          // background-color: #ffffff;
```
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+        position: relative;
+        padding: 0;
+        height: 8px;
+        border-color: #2389ff;
```
Should be declared as a var at the top of the file... also, should probably use an already defined variable and lighten or darken it? @dsmmcken

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+
+  border-radius: $border-radius;
+
+  // margin-bottom: 2px;
```
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.tsx:
```diff
@@ -0,0 +1,87 @@
+
+export const TreeItem = forwardRef<HTMLDivElement, Props>((props, ref) => {
+  const {
+    childCount = 0,
```
Unused prop... are you trying to assign default values to the props here? I think would need to do in the `renderItem` function defined in `VisibilityOrderingBuild` instead.

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.tsx:
```diff
@@ -0,0 +1,87 @@
+  } = props;
+
+  const DepthMarker = <span className={styles.DepthLine} />;
+  const depthMarkers = Array(depth).fill(DepthMarker);
```
`useMemo`

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/keyboardCoordinates.ts:
```diff
@@ -0,0 +1,158 @@
+  KeyboardCoordinateGetter,
+  DroppableContainer,
+} from '@dnd-kit/core';
+
```
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/keyboardCoordinates.ts:
```diff
@@ -0,0 +1,158 @@
+import type { SensorContext } from './types';
+import { getProjection } from './utilities';
+
+const directions: string[] = [
```
`SCREAMING_SNAKE_CASE`

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/keyboardCoordinates.ts:
```diff
@@ -0,0 +1,158 @@
+
+const horizontal: string[] = [KeyboardCode.Left, KeyboardCode.Right];
+
+export const sortableTreeKeyboardCoordinates: (
```
Why not declare as `function`? Also probably should have some tests for this function.

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/utilities.ts:
```diff
@@ -0,0 +1,345 @@
+import type { UniqueIdentifier } from '@dnd-kit/core';
```
I'd add some tests for some of the functions in this file.

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/utilities.ts:
```diff
@@ -0,0 +1,345 @@
+import type ColumnHeaderGroup from '../../../ColumnHeaderGroup';
+import type { FlattenedItem, TreeItem, TreeItems } from './types';
+
+export const iOS = /iPad|iPhone|iPod/.test(navigator.platform);
```
Should be named `isIOS`

mattrunyon replied:
Not needed since we don't do any mobile stuff

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -675,77 +617,333 @@ class VisibilityOrderingBuilder extends Component<
+    const columnGroupMap = model.getColumnHeaderGroupMap();
+    const newGroups = [...columnGroupMap.values()];
+
+    // TODO: Change to name based so you can seed with groups
+    // TODO: Remove children from their parents
```
Some TODOs. Noticed some errors:
- Creating a group when a group named `NewGroup` already exists
- Creating a group with child elements selected
- Creating a group with some top level and child elements selected

mattrunyon replied:
These should all be fixed now

mofojed commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/utilities.ts:
```diff
@@ -0,0 +1,345 @@
+  return items;
+}
+
+export function getProjection(
```
I'd add some docs on these utility functions

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -96,17 +107,18 @@ class VisibilityOrderingBuilder extends Component<
+    this.handleGroupDelete = this.handleGroupDelete.bind(this);
+    this.handleGroupNameChange = this.handleGroupNameChange.bind(this);
+    this.validateGroupName = this.validateGroupName.bind(this);
+    // this.handleGroupColorChange = this.handleGroupColorChange.bind(this);
```
```suggestion
```

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -142,15 +155,12 @@ class VisibilityOrderingBuilder extends Component<
     });
-    onMovedColumnsChanged([]);
+    onMovedColumnsChanged(model.movedColumns);
+    model.setColumnHeaderGroups(model.layoutHints?.columnGroups);
```
As mentioned elsewhere shouldn't need to call `setColumnHeaderGroups` from anywhere if you make it a `get`/`set` property and set it through `TableViewportUpdater`

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -675,77 +617,333 @@ class VisibilityOrderingBuilder extends Component<
+  makeVisibilityOrderingList = memoize(
+    (
+      movedColumns: MoveOperation[],
+      columnHeaderGroups: ColumnHeaderGroup[],
```
This is unused in the memoized function

mattrunyon replied:
I added a comment. It isn't used, but it affects the list when it changes. I think I need to adjust the `getTreeItems` util to take this in instead of grabbing it from the model

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -675,77 +617,333 @@ class VisibilityOrderingBuilder extends Component<
+      userColumnWidths: ModelSizeMap,
+      selectedColumns: ModelIndex[]
+    ) => {
+      const { model } = this.props;
+      const { columns } = model;
```
`columns` should be passed in as a param to the memoized function.

mofojed commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.tsx:
```diff
@@ -901,133 +1011,125 @@ class VisibilityOrderingBuilder extends Component<
+                : 'Create group'
+            }
+            disabled={selectedColumns.length === 0}
+            onClick={() => this.handleGroupCreate(selectedColumns)}
```
Not sure why you're doing this instead of just passing `onClick={this.handleGroupCreate}` and get `selectedColumns` from state in there.

mofojed reviewed (changes requested):
Also an `Uncaught Error: Column group child MyByte specified in multiple groups` if you click on a header that's in a group already, then click the `Group` button.

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+.Wrapper {
+  list-style: none;
+  box-sizing: border-box;
+  padding-left: var(--spacing);
```
Should be a $spacer-N?

mattrunyon replied:
Not used any more. Was initially the way the example set the indent for each item. Instead I added depth lines so this is now removed

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+        position: relative;
+        padding: 0;
+        height: 8px;
+        border-color: #2389ff;
+        background-color: #56a1f8;
+
+        &:before {
+          position: absolute;
+          left: -8px;
+          top: -4px;
+          display: block;
+          content: '';
+          width: 12px;
+          height: 12px;
+          border-radius: 50%;
+          border: 1px solid #2389ff;
```
😞 why all these hardcoded colours?

mattrunyon replied:
These aren't used. They're from the example which has an option for a smaller indicator of where you're dropping instead of a ghost item

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+.Count {
+  padding: 0 $spacer-2;
+  border-radius: 1rem;
+  background-color: rgba($white, 0.25);
```
$foreground?

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+
+.MarchingAnts {
+  @include ants-base($foreground, $background);
+  animation-name: march;
```
shouldn't need to include animation name

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+.DepthLine {
+  margin-left: $depth-margin;
+  padding-bottom: 2px;
+  height: 28px;
```
seems like a magic number, should probably be extracted to variable at the top, or calculated if it can be.

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+
+.Collapse {
+  svg {
+    transition: transform 250ms ease;
```
$transition-mid for time

mattrunyon replied:
Removed the collapse stuff since we aren't using that functionality from the example

dsmmcken commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/sortable-tree/TreeItem.module.scss:
```diff
@@ -0,0 +1,162 @@
+@import '@deephaven/components/scss/custom.scss';
+
+.Wrapper {
```
class names are normally kebab-case by convention. Did you do it this way because you are using it as a module?

mattrunyon replied:
I did it this way because I copied it from dnd-kit. I'll clean up to our conventions

mattrunyon commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.scss:
```diff
@@ -11,9 +11,25 @@
     z-index: 2;
+
+    .btn-link-icon {
+      padding-left: $spacer-1 !important;
+      padding-right: $spacer-1 !important;
```
Nope. Removing

mattrunyon commented on packages/iris-grid/src/sidebar/VisibilityOrderingBuilder.scss:
```diff
@@ -143,7 +159,7 @@
     padding-right: $spacer-1;
     margin-left: $spacer-1;
-    height: 1rem;
+    height: 70%;
```
Only reason I can remember changing this is `1rem` is really difficult to see the divider for me. It's also low contrast which doesn't help

mattrunyon commented on packages/iris-grid/src/IrisGridTableModelTemplate.ts:
```diff
@@ -170,7 +169,9 @@ class IrisGridTableModelTemplate<
   private _columnHeaderMaxDepth: number | null = null;
 
-  private _columnHeaderGroups: ColumnGroup[] = [];
+  private _columnHeaderGroups: ColumnHeaderGroup[] = [];
```
I'll remove it

mattrunyon commented on packages/iris-grid/src/sidebar/visibility-ordering-builder/VisibilityOrderingGroup.tsx:
```diff
@@ -0,0 +1,171 @@
+      }
+    };
+
+    return (
+      <div className="group-editing py-1">
+        <div style={{ display: 'flex', alignItems: 'center', flexGrow: 1 }}>
+          <input
+            ref={nameInputRef}
+            className={classNames('form-control', 'mr-1', {
+              'is-invalid': !isValid,
+            })}
+            style={{ flexGrow: 1 }}
+            value={name}
+            onChange={e => setName(e.target.value)}
+            onKeyDown={handleEditKeyDown}
+          />
+          <Button
+            kind="ghost"
+            className="p-1"
+            icon={vsCheck}
+            tooltip="Confirm"
+            onClick={handleConfirm}
+          />
+          <Button
+            kind="ghost"
+            className="p-1"
+            icon={vsChromeClose}
+            tooltip="Cancel"
+            onClick={handleCancel}
+          />
+        </div>
+        {!isValid && (
+          <p className="mb-0 validate-label-error text-danger">
+            Invalid or duplicate name
+          </p>
+        )}
+      </div>
+    );
```
Fair point. I don't think `useCallback` is necessary for the handle methods here though since they don't get passed down any further and will change on every update anyway. This only renders base elements so they'll all re-render if anything triggers a render. And controlled inputs will change the callback every time a key is pressed so memoizing them doesn't prevent re-renders

mattrunyon commented on packages/iris-grid/src/IrisGridUtils.ts:
```diff
@@ -338,8 +349,16 @@ class IrisGridUtils {
+
+    let hydratedColumnHeaderGroups: ColumnHeaderGroup[] | null = null;
+
+    if (columnHeaderGroups) {
```
Nope. If the user has deleted all `columnHeaderGroups`, it will be set to an empty array. The setter method handles empty arrays fine (as well as undefined for clearing). `null` (or undefined) is used for the "no overrides" value

mofojed replied:
Ah okay well then just `if (columnHeaderGroups != null) {`

I was more confused why the `strict-boolean-expressions` isn't tripping here, I thought it would require `!= null`

dsmmcken reviewed (changes requested):
Styling is looking pretty good. Getting closer.

Here's the things I noticed:

- [x] Bug: Toggle All doesn't toggle all to hidden when all our showing, only all to shown when one is hidden, that's a regression. When a selection is made the Hide Selected button doesn't work either
- [x] Bug: internal group name appears while it is first being created in table, before a name is even given :newgroup..
- [x] Create new group should be immediately cancelable triggering delete on first creation, otherwise you have to name it just to delete it
- [x] Rename section to "Hide, Group, and Order Columns"
- [x] Bug: Yes, animation is wonky with arrow moves, looks like it is always animating from [0,0]?
- [x] Bug: Create group with group as first item, select two other items further down the list and move them to top with arrow, then move down one with arrow. The first item entirely disappears from the list.
- [x] Bug: Moving a group down with the down arrow causes the group to be deleted (
- [x] Bug? Exiting with a group name in progress
- [x] Bug? Creating a new group, never naming it then close sidebar keeps it around in the initial edit state? Should it remove it?

mofojed commented:
There's one weird case where if you create a group, then don't type anything, then select a few others and create a group, you get two input fields, and in the grid they look like they're in the same (empty) group:
![image](https://user-images.githubusercontent.com/4505624/210378611-72b68774-6df5-42cc-ab85-2e279b423510.png)

I think if the input field blurs without any changes, should just cancel the grouping.

mofojed commented on packages/iris-grid/src/IrisGrid.tsx:
```diff
@@ -3885,8 +3899,10 @@ export class IrisGrid extends Component<IrisGridProps, IrisGridState> {
               model={model}
               movedColumns={movedColumns}
               userColumnWidths={userColumnWidths}
+              columnHeaderGroups={model.getColumnHeaderGroups()}
```
Hmmm why are you using the `model.getColumnHeaderGroups()` instead of the `columnHeaderGroups` from state?

mofojed commented on packages/iris-grid/src/IrisGridModel.ts:
```diff
@@ -487,6 +489,25 @@ abstract class IrisGridModel<
    */
   abstract delete(ranges: GridRange[]): Promise<void>;
+
+  getColumnHeaderGroups(): ColumnHeaderGroup[] {
+    return [];
+  }
+
+  getColumnHeaderGroupMap(): Map<string, ColumnHeaderGroup> {
+    return new Map();
+  }
+
+  setColumnHeaderGroups(groups: ColumnGroup[] | undefined): void {
+    // no-op
+  }
+
+  getColumnHeaderParentGroup(
+    modelIndex: ModelIndex,
+    depth: number
+  ): ColumnHeaderGroup | undefined {
+    return undefined;
+  }
```
This comment hasn't been addressed.

mofojed reviewed (changes requested):
Still need to address my IrisGridModel comments (`setColumnHeaderGroups` => `set columnHeaderGroups`, etc)

mattrunyon commented:
> There's one weird case where if you create a group, then don't type anything, then select a few others and create a group, you get two input fields, and in the grid they look like they're in the same (empty) group: ![image](https://user-images.githubusercontent.com/4505624/210378611-72b68774-6df5-42cc-ab85-2e279b423510.png)
> 
> I think if the input field blurs without any changes, should just cancel the grouping.

My concern with this is it makes it impossible to do something like create a group and then copy the group name from another source. Maybe the user is trying to clone a spreadsheet they have and want to copy the name.

I guess I could make it to where if another group is created with a new group that hasn't had its name changed, then the old new group would be deleted

mattrunyon commented on packages/components/src/popper/Popper.tsx:
```diff
@@ -150,6 +150,7 @@ class Popper extends Component<PopperProps, PopperState> {
     let { options } = this.props;
     options = {
       placement: 'auto',
+      modifiers: { preventOverflow: { boundariesElement: 'viewport' } },
```
It's part of that ticket. Binds to viewport. Didn't change the placement from auto though. Don wanted to check that it didn't cause overflow issues by changing default to bottom instead of auto

dsmmcken commented:
- [ ] Dragging a group to the top breaks ![uhoh](https://user-images.githubusercontent.com/1576283/210437325-e7a36293-8da4-4544-8e64-12dc37956b42.gif)
- [ ] Creating two unamed groups, then reset. It only removes one group and mylong, myfloat, mydouble are entirely missing from the list untile the second reset.
![uhoh2](https://user-images.githubusercontent.com/1576283/210437327-ccff239f-cd08-4033-93f7-c509961f751f.gif)

mattrunyon commented:
I changed to only allow 1 new group at a time which fixes the reset issue.

Fixed the issue of dragging a group to top. This only occurred if dragging from the top to the top.

mofojed commented on packages/iris-grid/src/IrisGridModel.ts:
```diff
@@ -133,7 +136,7 @@ abstract class IrisGridModel<
+  get initialMovedColumns(): MoveOperation[] {
     return [];
   }
 
```
`get movedRows()` should be `get initialMovedRows()` now to match new naming

mofojed commented on packages/iris-grid/src/IrisGrid.tsx:
```diff
@@ -3885,8 +3900,10 @@ export class IrisGrid extends Component<IrisGridProps, IrisGridState> {
               model={model}
               movedColumns={movedColumns}
               userColumnWidths={userColumnWidths}
+              columnHeaderGroups={model.columnHeaderGroups}
```
Can just be `columnHeaderGroups`

mofojed commented on packages/iris-grid/src/IrisGridModel.ts:
```diff
@@ -142,6 +145,113 @@ abstract class IrisGridModel<
+   *
+   * @returns Object containing groups array, max depth, map of name to parent group, and map of name to group
+   */
+  parseColumnHeaderGroups = memoize(
```
Pull this into `IrisGridUtils` as a utility function. Memoization shouldn't be needed in `IrisGridUtils` or `IrisGridTableModelTemplate`; `VisibilityOrderingBuilder` should be calling something like `get initialColumnHeaderGroups` instead.

mofojed commented on packages/iris-grid/src/IrisGridTableModelTemplate.ts:
```diff
@@ -56,6 +56,8 @@ const log = Log.module('IrisGridTableModel');
 const SET_VIEWPORT_THROTTLE = 150;
 const APPLY_VIEWPORT_THROTTLE = 0;
 
+const EMPTY_ARRAY: unknown[] = [];
```
Just making a note that I'll replace this in my readonly change: https://github.com/deephaven/web-client-ui/pull/986

mofojed reviewed (dismissed):
Side note, any reason we don't allow changing the colour on columns individually and limit it to just groups? Seems like that would be nice functionality to have @dsmmcken

dsmmcken commented:
> any reason we don't allow changing the colour on columns individually and limit it to just groups

Yeah, we could consider that for a future revision. And if we are colouring columns, we could also maybe consider having columns inherit their parents colour?
